src/script.c

```diff
@@ -141,15 +145,234 @@ void environment_exit(environment_t *env) {
+			logger(DEBUG_ALWAYS, LOG_ERR, "Got empty or invalid header: %s", strerror(errno));
+			return;
+		}
+
+		// Reset all known environment variables
+		for(const char **known = known_vars; *known; ++known) {
+			unsetenv(*known);
+		}
+
+		size_t len = 0;
+
+		for(int i = 0; i < hdr.vars; ++i) {
+			// Read environment variable
+			if(!read_unix(sock, &len, sizeof(len)) || len >= sizeof(buf) || !read_unix(sock, buf, len)) {
+				logger(DEBUG_ALWAYS, LOG_ERR, "Got incorrect environment variable value: %s", strerror(errno));
+				return;
+			}
+
+			buf[len] = '\0';
+
+			// Check that we received a valid env var expression: NAME=value
+			char *var_val = strchr(buf, '=');
+
+			if(!var_val || var_val == buf) {
+				logger(DEBUG_ALWAYS, LOG_EMERG, "Got broken environment variable (possible attack)");
+				return;
+			}
+
+			// Replace '=' with '\0', splitting env var expression into name and value
+			*var_val = '\0';
+
+			// Check that tincd didn't pass anything weird or dangerous (like LD_PRELOAD)
+			if(!is_known_var(buf)) {
+				logger(DEBUG_ALWAYS, LOG_EMERG, "Got unknown environment variable (possible attack)");
+				return;
+			}
+
+			setenv(buf, var_val + 1, 1);
+		}
+
+		// Read script name
+		if(!read_unix(sock, &len, sizeof(len)) || len >= sizeof(buf) || !read_unix(sock, buf, len)) {
+			logger(DEBUG_ALWAYS, LOG_ERR, "Got incorrect command: %s", strerror(errno));
+			return;
+		}
+
+		buf[len] = '\0';
+
+		char scriptname[PATH_MAX];
+
+		// Check it for signs of possible attack (basically paths to anything other
+		// than tincd scripts in its configuration directory), and build full command.
+		if(!build_script_name(scriptname, sizeof(scriptname), buf)) {
+			logger(DEBUG_ALWAYS, LOG_EMERG, "Got invalid script name (possible attack)");
+			return;
+		}
+
+		const cmd_result_t result = {
+			.magic = CMD_MAGIC,
+			.current = hdr.current,
+			.status = run_command(scriptname),
+		};
+
+		// Send current iteration number and command exit status to tincd
+		if(!write_unix(sock, &result, sizeof(result))) {
+			logger(DEBUG_ALWAYS, LOG_ERR, "Could not write command result: %s", strerror(errno));
+			return;
+		}
+	}
+}
+
+void run_script_worker(uid_t uid, int sock) {
+	if(uid && setuid(uid)) {
+		fprintf(stderr, "Could not set user ID %d\n", uid);
+		exit(EXIT_FAILURE);
+	}
+
+#ifdef HAVE_LINUX
+	prctl(PR_SET_NAME, "scripts");
+#endif
+
+	script_worker_loop(sock);
+	close(sock);
+
+	logger(DEBUG_ALWAYS, LOG_NOTICE, "Script worker is terminating");
+	exit(EXIT_SUCCESS);
+}
+#endif // HAVE_SANDBOX
+
 bool execute_script(const char *name, environment_t *env) {
-	if(!sandbox_can(START_PROCESSES, RIGHT_NOW)) {
+#ifdef HAVE_SANDBOX
+
+	if(!sandbox_can(RUN_SCRIPTS, RIGHT_NOW)) {
 		return false;
 	}
 
+	if(worker.stopped) {
+		logger(DEBUG_ALWAYS, LOG_ERR, "Script worker is broken, scripts will not work until tincd is restarted");
+		return false;
+	}
```

Comment:
I wonder what we should do here. It's one thing that a script itself fails, but if the script worker fails we won't even try to run the scripts themselves anymore. If someone is relying on something in a script, like say setting firewall rules to allow or block traffic from a node that just came online, this will be problematic. Perhaps it's better to just terminate tincd altogether when that happens.

Reply:
Good idea, this gives the supervisor (like systemd) an opportunity to 'fix' the issue.